bookwyrm/migrations/0179_populate_sort_title.py

```diff
@@ -0,0 +1,41 @@
+import re
+from itertools import chain
+
+from django.db import migrations, transaction
+from django.db.models import Q
+
+from bookwyrm.settings import LANGUAGE_ARTICLES
+
+
+@transaction.atomic
+def populate_sort_title(apps, schema_editor):
```

Comment:
I think this will work well for smaller instances, but for bigger ones like bookwyrm.social, it's going to be too intensive to iterate over all editions. This could work as a management command which admins can run optionally, however. What do you think?

Reply:
Do you think something like [`bulk_update`](https://docs.djangoproject.com/en/4.2/ref/models/querysets/#bulk-update) would work for larger instances?

Reply:
That's a good question! I don't know -- on b.s, there are about 700,000 editions, if that helps contextualize it. @WesleyAC do you have a sense?

Reply:
There are probably some edge cases were "A" is used as a noun in the title such as [A is for Activist](https://bookwyrm.social/book/257447/s/a-is-for-activist).  We could probably get fancy and exclude titles that start with "A is" or even fancier with "A [verb]", but then we open a can of worms implementing custom rules for every language.  Perhaps the planned addition of the ability to edit the sort title on the book edit page mitigates this concern?

Reply:
Thinking about this some more I added some logic to slice the queryset using batches.  This way we don't try to hold all the editions in memory at the same time.  The `bulk_update` should limit the # of queries.  I would love to get additional thoughts on this and whether it would work.

```python
    editions_wo_sort_title = Edition.objects.using(db_alias).filter(
        Q(sort_title__isnull=True) | Q(sort_title__exact="")
    )
    batch_size = 50000
    start = 0
    end = batch_size
    while editions_wo_sort_title[start: end]:
        Edition.objects.bulk_update((
            set_sort_title(edition)
            for edition in editions_wo_sort_title[start: end]
        ), ["sort_title"])
        start = end
        end += batch_size
```

Reply:
@zachflanders if it's something you're comfortable with, could you change the postgres `log_statement` setting to `all` and paste the postgres logs from running this migration here (or probably a link to a gist)? that would help in determining how impactful it's likely to be.

if not, i can take a look at it, but it might take me a little while to get to it.

Reply:
@mouse-reeve @WesleyAC With a little help from [this SO answer](https://stackoverflow.com/a/71270291/3669465) I refactored this a bit.  Here is a link [a gist of the database logs](https://gist.github.com/zachflanders/d14a9035273626cf90d1dbbe1cb91237).  I have 522 editions in my local dev environment, and I lowered the batch size to 20 in order to mimic the number of iterations a large instance like BS would likely have. (Of course each iteration will update many more rows, but it would be the same number of queries)